File: RPi_motor_interface/src/initial_pose_publisher.cpp
#include "motor_interface/initial_pose_publisher.hpp"

InitialPosePublisher::InitialPosePublisher()
: Node("initial_pose_publisher"), published_(false)
{
  publisher_ = this->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("/initialpose", 10);
  timer_ = this->create_wall_timer(
    std::chrono::seconds(1),
    std::bind(&InitialPosePublisher::publish_initial_pose, this)
  );
}

void InitialPosePublisher::publish_initial_pose()
{
  if (published_) 
  {
    return;
  }

  geometry_msgs::msg::PoseWithCovarianceStamped msg;
  msg.header.stamp = this->get_clock()->now();
  msg.header.frame_id = "map";

  msg.pose.pose.position.x = 0.0;
  msg.pose.pose.position.y = 0.0;
  msg.pose.pose.orientation.w = 1.0;

  msg.pose.covariance = {
    0.25, 0,    0, 0, 0, 0,
    0,    0.25, 0, 0, 0, 0,
    0,    0,    0, 0, 0, 0,
    0,    0,    0, 0, 0, 0,
    0,    0,    0, 0, 0, 0,
    0,    0,    0, 0, 0, 0.0685
  };

  publisher_->publish(msg);
  RCLCPP_INFO(this->get_logger(), "Initial pose published.");
  published_ = true;
}

// main
int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<InitialPosePublisher>());
  rclcpp::shutdown();
  return 0;
}

File: RPi_motor_interface/obsolete/initial_pose_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseWithCovarianceStamped

class InitialPosePublisher(Node):
    def __init__(self):
        super().__init__('initial_pose_publisher')
        self.publisher_ = self.create_publisher(PoseWithCovarianceStamped, '/initialpose', 10)
        self.timer = self.create_timer(1.0, self.publish_initial_pose)
        self.published = False

    def publish_initial_pose(self):
        if self.published:
            return

        msg = PoseWithCovarianceStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'map'
        msg.pose.pose.position.x = 0.0
        msg.pose.pose.position.y = 0.0
        msg.pose.pose.orientation.w = 1.0

        msg.pose.covariance = [
            0.25, 0, 0, 0, 0, 0,
            0, 0.25, 0, 0, 0, 0,
            0, 0, 0.0, 0, 0, 0,
            0, 0, 0, 0.0, 0, 0,
            0, 0, 0, 0, 0.0, 0,
            0, 0, 0, 0, 0, 0.06853891945200942
        ]

        self.publisher_.publish(msg)
        self.get_logger().info('Initial pose published.')
        self.published = True


def main(args=None):
    rclpy.init(args=args)
    node = InitialPosePublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

